From e1a585d37d41dfc8c3619295bfe09adb4c262c77 Mon Sep 17 00:00:00 2001
From: xianhaox <xianhaox.qing@intel.com>
Date: Tue, 16 Jan 2018 17:08:26 +0800
Subject: [PATCH] Filter null file path avoid crash

VolumeInfo object return from StorageManager may contains user cannot
access to which will cause null path on calling vol.getPathForUser().
The null path will be passed to the ListView adapter later.
Once user clicked on the item with null path the app will crash.

The solution is using another API getVolumeList() to get volumes.
getVolumeList() will filter the volume internally.

tests:
adb shell monkey --pct-syskeys 0 -p com.android.car.systemupdater 1000
The SystemUpdater app should not crash after 1000 events injected.

Change-Id: I8744c0001a30b8007feb0523b9003cd3f06d9605
Tracked-On: https://jira01.devtools.intel.com/browse/OAM-55877
Signed-off-by: xianhaox <xianhaox.qing@intel.com>
---
 .../android/car/systemupdater/SystemUpdaterActivity.java    | 13 +++++--------
 1 file changed, 5 insertions(+), 8 deletions(-)

diff --git a/src/com/android/car/systemupdater/SystemUpdaterActivity.java b/src/com/android/car/systemupdater/SystemUpdaterActivity.java
index e442731..7234df9 100644
--- a/src/com/android/car/systemupdater/SystemUpdaterActivity.java
+++ b/src/com/android/car/systemupdater/SystemUpdaterActivity.java
@@ -40,6 +40,7 @@ import java.util.List;
 
 import android.os.storage.StorageEventListener;
 import android.os.storage.StorageManager;
+import android.os.storage.StorageVolume;
 import android.os.storage.VolumeInfo;
 
 /**
@@ -108,15 +109,11 @@ public class SystemUpdaterActivity extends Activity {
         if (mStorageManager == null) {
             return;
         }
-        final List<VolumeInfo> vols = mStorageManager.getVolumes();
-        File[] files = new File[vols.size()];
+        final StorageVolume[] vols = mStorageManager.getVolumeList();
+        File[] files = new File[vols.length];
         int i = 0;
-        for (VolumeInfo vol : vols) {
-            File path = vol.getPathForUser(getUserId());
-            if (vol.getState() != VolumeInfo.STATE_MOUNTED || path == null) {
-                continue;
-            }
-            files[i++] = path;
+        for (StorageVolume vol : vols) {
+            files[i++] = vol.getPathFile();
         }
         getFragmentManager().popBackStack(null, FragmentManager.POP_BACK_STACK_INCLUSIVE);
         DeviceListFragment frag = new DeviceListFragment();
-- 
1.9.1

